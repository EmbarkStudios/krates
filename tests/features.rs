mod util;

/// Ensures weak dependencies are properly pruned if not explicitly pulled in
/// https://github.com/EmbarkStudios/krates/issues/41
#[test]
fn prunes_multiple_weak_features() {
    let mut cmd = krates::Cmd::new();
    cmd.manifest_path("tests/features/Cargo.toml")
        .features(
            ["blocking", "json", "multipart", "stream"]
                .into_iter()
                .map(|f| f.to_owned()),
        )
        .no_default_features();

    let mut builder = krates::Builder::new();
    builder.include_targets([("x86_64-unknown-linux-gnu", vec![])]);
    let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    for name in md.krates().map(|k| &k.name) {
        println!("{name}");
    }

    // All of the specified features have a weak dependency on reqwest, so it
    // shouldn't be present in the graph
    assert_eq!(0, md.krates_by_name("reqwest").count());
    assert_eq!(0, md.krates_by_name("reqest").count());
}

/// While the zlib features brings in git2, the openssl dependency for both
/// git2 and git2-sys is optional and weak, and since we've not explicitly
/// enabled a feature to bring it in, it should not be present in the graph,
/// even though `cargo metadata` will list it in the graph
#[test]
fn prunes_mixed_dependencies() {
    let mut cmd = krates::Cmd::new();
    cmd.manifest_path("tests/features/Cargo.toml")
        .features(["zlib".to_owned()]);

    let mut builder = krates::Builder::new();
    builder.include_targets([("x86_64-unknown-linux-gnu", vec![])]);
    let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    assert_eq!(0, md.krates_by_name("openssl-sys").count());
    // cmake is brought in via the zlib-ng-compat feature. gross.
    assert_eq!(1, md.krates_by_name("cmake").count());
}

macro_rules! assert_features {
    ($graph:expr, $name:expr, $features:expr) => {
        let (_, krate) = $graph.krates_by_name($name).next().unwrap();

        assert_eq!($graph.get_enabled_features(&krate.id).unwrap(), $features);
    };
}

/// Ensures we can enable crate features even when that crate has been renamed
/// as well as features being precise
#[test]
fn handles_features_for_renamed_crates() {
    let mut cmd = krates::Cmd::new();
    cmd.manifest_path("tests/features/Cargo.toml")
        .no_default_features()
        .features(["midi".to_owned()]);

    let mut builder = krates::Builder::new();
    builder.include_targets([("aarch64-apple-darwin", vec![])]);
    let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    assert_eq!(1, md.krates_by_name("coreaudio-rs").count());

    // Ensure only the single feature is enabled, not all of them
    assert_features!(md, "coreaudio-sys", ["core_midi"]);
}

/// Ensures that explicitly toggling on an optional dependency works
#[test]
fn handles_explicit_weak_features() {
    {
        let mut cmd = krates::Cmd::new();
        cmd.manifest_path("tests/features/Cargo.toml")
            .no_default_features()
            .features(["reqest".to_owned()]);

        let mut builder = krates::Builder::new();
        builder.include_targets([("x86_64-unknown-linux-musl", vec![])]);
        let md: util::Graph = builder.build(cmd, krates::NoneFilter).unwrap();

        let dg = krates::petgraph::dot::Dot::new(md.graph()).to_string();

        std::fs::write("ohno.dot", dg).unwrap();

        // assert_features!(
        //     md,
        //     "reqwest",
        //     [
        //         "__rustls",
        //         "__tls",
        //         "async-compression",
        //         "brotli",
        //         "cookie_crate",
        //         "cookie_store",
        //         "cookies",
        //         "hyper-rustls",
        //         "proc-macro-hack",
        //         "rustls",
        //         "rustls-pemfile",
        //         "rustls-tls",
        //         "rustls-tls-webpki-roots",
        //         "tokio-rustls",
        //         "tokio-util",
        //         "webpki-roots"
        //     ]
        // );
    }

    // {
    //     let mut cmd = krates::Cmd::new();
    //     cmd.manifest_path("tests/features/Cargo.toml")
    //         .no_default_features()
    //         .features(["reqest".to_owned()]);

    //     let mut builder = krates::Builder::new();
    //     builder.include_targets([("x86_64-unknown-linux-musl", vec![])]);
    //     builder.ignore_kind(krates::DepKind::Normal, krates::Scope::All);
    //     let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    //     assert_features!(
    //         md,
    //         "reqwest",
    //         ["cookie_crate", "cookie_store", "cookies", "proc-macro-hack"]
    //     );
    // }

    // let mut cmd = krates::Cmd::new();
    // cmd.manifest_path("tests/features/Cargo.toml")
    //     .features(["reqest".to_owned()]);

    // let mut builder = krates::Builder::new();
    // builder.include_targets([("x86_64-unknown-linux-musl", vec![])]);
    // let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    // assert_features!(md, "reqwest", ["json", "cookies"]);
}
