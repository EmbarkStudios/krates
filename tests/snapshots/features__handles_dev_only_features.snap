---
source: tests/features.rs
expression: "krates::petgraph::dot::Dot::new(md.graph()).to_string()"
---
digraph {
    0 [ label = "crate base-x 0.2.11" ]
    1 [ label = "crate cbor4ii 0.2.14" ]
    2 [ label = "crate cid 0.10.1" ]
    3 [ label = "crate core2 0.4.0" ]
    4 [ label = "crate data-encoding 2.5.0" ]
    5 [ label = "crate data-encoding-macro 0.1.14" ]
    6 [ label = "crate data-encoding-macro-internal 0.1.12" ]
    7 [ label = "crate features-galore 0.1.0 path+file:///krates/tests/features" ]
    8 [ label = "crate memchr 2.5.0" ]
    9 [ label = "crate multibase 0.9.1" ]
    10 [ label = "crate multihash 0.18.1" ]
    11 [ label = "crate multihash-derive 0.8.1" ]
    12 [ label = "crate proc-macro-crate 1.1.3" ]
    13 [ label = "crate proc-macro-error 1.0.4" ]
    14 [ label = "crate proc-macro-error-attr 1.0.4" ]
    15 [ label = "crate proc-macro2 1.0.76" ]
    16 [ label = "crate quote 1.0.35" ]
    17 [ label = "crate scopeguard 1.2.0" ]
    18 [ label = "crate serde 1.0.195" ]
    19 [ label = "crate serde-big-array 0.3.3" ]
    20 [ label = "crate serde_bytes 0.11.14" ]
    21 [ label = "crate serde_derive 1.0.195" ]
    22 [ label = "crate serde_ipld_dagcbor 0.4.2" ]
    23 [ label = "crate syn 1.0.98" ]
    24 [ label = "crate syn 2.0.48" ]
    25 [ label = "crate synstructure 0.12.6" ]
    26 [ label = "crate thiserror 1.0.56" ]
    27 [ label = "crate thiserror-impl 1.0.56" ]
    28 [ label = "crate toml 0.5.11" ]
    29 [ label = "crate unicode-ident 1.0.1" ]
    30 [ label = "crate unicode-xid 0.2.4" ]
    31 [ label = "crate unsigned-varint 0.7.2" ]
    32 [ label = "crate version_check 0.9.4" ]
    33 [ label = "feature std" ]
    34 [ label = "feature alloc" ]
    35 [ label = "feature std" ]
    36 [ label = "feature alloc" ]
    37 [ label = "feature serde-codec" ]
    38 [ label = "feature std" ]
    39 [ label = "feature alloc" ]
    40 [ label = "feature default" ]
    41 [ label = "feature std" ]
    42 [ label = "feature alloc" ]
    43 [ label = "feature parsing" ]
    44 [ label = "feature proc-macro" ]
    45 [ label = "feature default" ]
    46 [ label = "feature std" ]
    47 [ label = "feature std" ]
    48 [ label = "feature derive" ]
    49 [ label = "feature const-generics" ]
    50 [ label = "feature default" ]
    51 [ label = "feature span-locations" ]
    52 [ label = "feature default" ]
    53 [ label = "feature default" ]
    54 [ label = "feature default" ]
    55 [ label = "feature default" ]
    56 [ label = "feature default" ]
    57 [ label = "feature proc-macro" ]
    58 [ label = "feature default" ]
    59 [ label = "feature default" ]
    60 [ label = "feature use_alloc" ]
    61 [ label = "feature use_std" ]
    62 [ label = "feature serde-codec" ]
    63 [ label = "feature std" ]
    64 [ label = "feature default" ]
    65 [ label = "feature proc-macro" ]
    66 [ label = "feature derive" ]
    67 [ label = "feature printing" ]
    68 [ label = "feature clone-impls" ]
    69 [ label = "feature visit" ]
    70 [ label = "feature extra-traits" ]
    71 [ label = "feature default" ]
    72 [ label = "feature default" ]
    73 [ label = "feature alloc" ]
    74 [ label = "feature serde_bytes" ]
    75 [ label = "feature serde" ]
    76 [ label = "feature multibase" ]
    77 [ label = "feature serde" ]
    78 [ label = "feature serde-big-array" ]
    79 [ label = "feature syn-error" ]
    80 [ label = "feature syn" ]
    81 [ label = "feature use_std" ]
    82 [ label = "feature serde_derive" ]
    83 [ label = "feature std" ]
    84 [ label = "feature std" ]
    85 [ label = "feature quote" ]
    86 [ label = "feature quote" ]
    87 [ label = "feature proc-macro" ]
    88 [ label = "feature printing" ]
    89 [ label = "feature parsing" ]
    90 [ label = "feature derive" ]
    91 [ label = "feature clone-impls" ]
    92 [ label = "feature proc-macro" ]
    1 -> 18 [ label = "" ]
    1 -> 33 [ label = "" ]
    2 -> 3 [ label = "" ]
    2 -> 34 [ label = "" ]
    2 -> 9 [ label = "" ]
    2 -> 35 [ label = "" ]
    2 -> 10 [ label = "" ]
    2 -> 36 [ label = "" ]
    2 -> 37 [ label = "" ]
    2 -> 38 [ label = "" ]
    2 -> 18 [ label = "" ]
    2 -> 39 [ label = "" ]
    2 -> 33 [ label = "" ]
    2 -> 40 [ label = "" ]
    2 -> 31 [ label = "" ]
    2 -> 41 [ label = "" ]
    3 -> 8 [ label = "" ]
    5 -> 4 [ label = "" ]
    5 -> 6 [ label = "" ]
    6 -> 42 [ label = "" ]
    6 -> 43 [ label = "" ]
    6 -> 44 [ label = "" ]
    7 -> 45 [ label = " 'cfg(windows)'" ]
    9 -> 0 [ label = "" ]
    9 -> 42 [ label = "" ]
    9 -> 46 [ label = "" ]
    9 -> 5 [ label = "" ]
    10 -> 3 [ label = "" ]
    10 -> 34 [ label = "" ]
    10 -> 11 [ label = "" ]
    10 -> 47 [ label = "" ]
    10 -> 48 [ label = "" ]
    10 -> 49 [ label = "" ]
    10 -> 31 [ label = "" ]
    10 -> 41 [ label = "" ]
    11 -> 12 [ label = "" ]
    11 -> 50 [ label = "" ]
    11 -> 51 [ label = "" ]
    11 -> 52 [ label = "" ]
    11 -> 53 [ label = "" ]
    11 -> 54 [ label = "" ]
    11 -> 55 [ label = "" ]
    12 -> 26 [ label = "" ]
    12 -> 56 [ label = "" ]
    13 -> 14 [ label = "" ]
    13 -> 52 [ label = "" ]
    13 -> 53 [ label = "" ]
    13 -> 23 [ label = "" ]
    13 -> 32 [ label = "(build)" ]
    14 -> 52 [ label = "" ]
    14 -> 53 [ label = "" ]
    14 -> 32 [ label = "(build)" ]
    15 -> 29 [ label = "" ]
    16 -> 15 [ label = "" ]
    16 -> 57 [ label = "" ]
    18 -> 58 [ label = "" ]
    19 -> 18 [ label = "" ]
    20 -> 18 [ label = "" ]
    20 -> 33 [ label = "" ]
    21 -> 52 [ label = "" ]
    21 -> 53 [ label = "" ]
    21 -> 59 [ label = "" ]
    22 -> 60 [ label = "" ]
    22 -> 61 [ label = "" ]
    22 -> 62 [ label = "" ]
    22 -> 63 [ label = "" ]
    22 -> 64 [ label = "" ]
    22 -> 39 [ label = "" ]
    22 -> 33 [ label = "" ]
    23 -> 15 [ label = "" ]
    23 -> 57 [ label = "" ]
    23 -> 16 [ label = "" ]
    23 -> 65 [ label = "" ]
    23 -> 29 [ label = "" ]
    24 -> 15 [ label = "" ]
    24 -> 57 [ label = "" ]
    24 -> 16 [ label = "" ]
    24 -> 65 [ label = "" ]
    24 -> 29 [ label = "" ]
    25 -> 15 [ label = "" ]
    25 -> 57 [ label = "" ]
    25 -> 16 [ label = "" ]
    25 -> 65 [ label = "" ]
    25 -> 66 [ label = "" ]
    25 -> 43 [ label = "" ]
    25 -> 67 [ label = "" ]
    25 -> 68 [ label = "" ]
    25 -> 69 [ label = "" ]
    25 -> 70 [ label = "" ]
    25 -> 44 [ label = "" ]
    25 -> 71 [ label = "" ]
    26 -> 27 [ label = "" ]
    27 -> 52 [ label = "" ]
    27 -> 53 [ label = "" ]
    27 -> 59 [ label = "" ]
    28 -> 72 [ label = "" ]
    61 -> 1 [ label = "" ]
    61 -> 33 [ label = "" ]
    61 -> 60 [ label = "" ]
    60 -> 1 [ label = "" ]
    63 -> 2 [ label = "" ]
    63 -> 38 [ label = "" ]
    63 -> 41 [ label = "" ]
    63 -> 73 [ label = "" ]
    63 -> 35 [ label = "" ]
    63 -> 33 [ label = "" ]
    74 -> 2 [ label = "" ]
    74 -> 20 [ label = "" ]
    62 -> 2 [ label = "" ]
    62 -> 73 [ label = "" ]
    62 -> 75 [ label = "" ]
    62 -> 37 [ label = "" ]
    62 -> 74 [ label = "" ]
    75 -> 2 [ label = "" ]
    75 -> 18 [ label = "" ]
    76 -> 2 [ label = "" ]
    76 -> 9 [ label = "" ]
    73 -> 2 [ label = "" ]
    73 -> 76 [ label = "" ]
    73 -> 36 [ label = "" ]
    73 -> 34 [ label = "" ]
    73 -> 39 [ label = "" ]
    34 -> 3 [ label = "" ]
    46 -> 4 [ label = "" ]
    46 -> 42 [ label = "" ]
    42 -> 4 [ label = "" ]
    35 -> 9 [ label = "" ]
    35 -> 46 [ label = "" ]
    38 -> 10 [ label = "" ]
    38 -> 41 [ label = "" ]
    38 -> 47 [ label = "" ]
    38 -> 36 [ label = "" ]
    37 -> 10 [ label = "" ]
    37 -> 77 [ label = "" ]
    37 -> 78 [ label = "" ]
    78 -> 10 [ label = "" ]
    78 -> 19 [ label = "" ]
    77 -> 10 [ label = "" ]
    77 -> 18 [ label = "" ]
    36 -> 10 [ label = "" ]
    36 -> 34 [ label = "" ]
    47 -> 11 [ label = "" ]
    79 -> 13 [ label = "" ]
    79 -> 80 [ label = "" ]
    80 -> 13 [ label = "" ]
    80 -> 23 [ label = "" ]
    50 -> 13 [ label = "" ]
    50 -> 79 [ label = "" ]
    51 -> 15 [ label = "" ]
    57 -> 15 [ label = "" ]
    52 -> 15 [ label = "" ]
    52 -> 57 [ label = "" ]
    65 -> 16 [ label = "" ]
    65 -> 57 [ label = "" ]
    53 -> 16 [ label = "" ]
    53 -> 65 [ label = "" ]
    81 -> 17 [ label = "" ]
    64 -> 17 [ label = "" ]
    64 -> 81 [ label = "" ]
    33 -> 18 [ label = "" ]
    82 -> 18 [ label = "" ]
    82 -> 21 [ label = "" ]
    48 -> 18 [ label = "" ]
    48 -> 82 [ label = "" ]
    72 -> 18 [ label = "" ]
    72 -> 33 [ label = "" ]
    39 -> 18 [ label = "" ]
    49 -> 19 [ label = "" ]
    83 -> 20 [ label = "" ]
    83 -> 33 [ label = "" ]
    40 -> 20 [ label = "" ]
    40 -> 83 [ label = "" ]
    58 -> 21 [ label = "" ]
    84 -> 22 [ label = "" ]
    84 -> 61 [ label = "" ]
    84 -> 63 [ label = "" ]
    84 -> 33 [ label = "" ]
    45 -> 22 [ label = "" ]
    45 -> 84 [ label = "" ]
    69 -> 23 [ label = "" ]
    85 -> 23 [ label = "" ]
    85 -> 16 [ label = "" ]
    44 -> 23 [ label = "" ]
    44 -> 57 [ label = "" ]
    44 -> 65 [ label = "" ]
    67 -> 23 [ label = "" ]
    67 -> 85 [ label = "" ]
    43 -> 23 [ label = "" ]
    70 -> 23 [ label = "" ]
    66 -> 23 [ label = "" ]
    54 -> 23 [ label = "" ]
    54 -> 66 [ label = "" ]
    54 -> 43 [ label = "" ]
    54 -> 67 [ label = "" ]
    54 -> 68 [ label = "" ]
    54 -> 44 [ label = "" ]
    68 -> 23 [ label = "" ]
    86 -> 24 [ label = "" ]
    86 -> 16 [ label = "" ]
    87 -> 24 [ label = "" ]
    87 -> 57 [ label = "" ]
    87 -> 65 [ label = "" ]
    88 -> 24 [ label = "" ]
    88 -> 86 [ label = "" ]
    89 -> 24 [ label = "" ]
    90 -> 24 [ label = "" ]
    59 -> 24 [ label = "" ]
    59 -> 90 [ label = "" ]
    59 -> 89 [ label = "" ]
    59 -> 88 [ label = "" ]
    59 -> 91 [ label = "" ]
    59 -> 87 [ label = "" ]
    91 -> 24 [ label = "" ]
    92 -> 25 [ label = "" ]
    92 -> 57 [ label = "" ]
    92 -> 44 [ label = "" ]
    92 -> 65 [ label = "" ]
    55 -> 25 [ label = "" ]
    55 -> 92 [ label = "" ]
    56 -> 28 [ label = "" ]
    71 -> 30 [ label = "" ]
    41 -> 31 [ label = "" ]
}

